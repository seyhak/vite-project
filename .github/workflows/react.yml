name: React CI

on:
  push:
    branches: [ "master" ]
  pull_request:
    branches: [ "master" ]

jobs:
  build:
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v3
    - name: Set up Node 18.18
      uses: actions/setup-node@v4
      with:
        node-version: 18.18
    - name: Install Dependencies
      run: |
        yarn
    # - name: Run Tests TZ issues
    #   run: |
    #     yarn test
    - name: Run Linters
      run: |
        yarn lint
