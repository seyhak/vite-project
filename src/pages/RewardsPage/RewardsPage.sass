.RewardsPage
  display: flex
  justify-content: center
  flex-direction: column
  text-align: center
  .card
    background-color: #1d1f2b
    padding: 1rem
    margin: 1rem
    min-width: 400px
  .total
    box-shadow: 0 0 10px blue
  .rewards
    .rewards-group
      .month-rewards-wrapper
        margin: 1rem 0
      & > h4
        margin: 0

