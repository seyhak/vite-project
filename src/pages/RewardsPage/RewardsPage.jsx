import { Loader } from "@components/Loader/Loader"
import { Reward } from "@components/Reward/Reward"
import { useRewardsPage } from "./useRewardsPage"

import "./RewardsPage.sass"

export const RewardsPage = () => {
  const { user, loading, rewards, total } = useRewardsPage()

  return (
    <div className="RewardsPage">
      {!user || loading || !rewards ? (
        <Loader />
      ) : (
        <>
          <h3>{`Welcome ${user.name}!`}</h3>
          <div className="rewards">
            {rewards
              ? Object.entries(rewards).map(([month, value]) => (
                  <div key={month} className="card rewards-group">
                    <h4 className="month">{month}</h4>
                    <div className="month-rewards-wrapper">
                      {value.transactions.length
                        ? value.transactions.map((reward) => (
                            <Reward
                              key={reward.date.getTime()}
                              value={reward.value}
                              rewardValue={reward.rewards}
                              date={reward.date}
                            />
                          ))
                        : "No transactions :C"}
                    </div>
                    <h4 className="sum">
                      {`Your reward points for the month: ${value.rewardsSum}`}
                    </h4>
                  </div>
                ))
              : "You have no rewards"}
          </div>
          <div className="total card">
            <h4 className="sum">{`Total points: ${total}`}</h4>
          </div>
        </>
      )}
    </div>
  )
}
