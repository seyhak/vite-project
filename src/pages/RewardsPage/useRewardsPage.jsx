import { useCallback, useEffect, useMemo } from "react"
import { useUser } from "@hooks/useUser"
import { useNavigate } from "react-router-dom"
import { PAGE_ROUTES } from "@constants/routes"
import { useFetch } from "@hooks/useFetch"
import { TRANSACTIONS_GET_URL } from "@constants/routes"
import { processTransactionsIntoRewards } from "@src/utils/transactions-to-rewards-processor"
import "./RewardsPage.sass"

export const useRewardsPage = () => {
  const navigate = useNavigate()
  const { user } = useUser()
  const {
    data,
    loading,
    error,
    refetch: getRewards,
  } = useFetch(TRANSACTIONS_GET_URL, true)

  const handleGetRewards = useCallback(async () => {
    await getRewards()
  }, [getRewards])

  useEffect(() => {
    if (!user) {
      navigate(PAGE_ROUTES.LOGIN)
    } else {
      handleGetRewards()
    }
  }, [user, handleGetRewards, navigate])

  const rewards = useMemo(() => {
    return data ? processTransactionsIntoRewards(data) : null
  }, [data])

  const total = useMemo(() => {
    return rewards
      ? Object.values(rewards).reduce((acc, curr) => {
          return acc + curr.rewardsSum
        }, 0)
      : 0
  }, [rewards])

  return { user, loading, rewards: rewards, error, total }
}
