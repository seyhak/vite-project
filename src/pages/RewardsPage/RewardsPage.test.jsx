import { describe, it, beforeEach, expect, vi } from "vitest"
import { render, screen } from "@testing-library/react"
import { RewardsPage } from "./RewardsPage"

const mockData = {
  January: {
    transactions: [
      { date: new Date("2013-01-05T23:00:00.000Z"), value: 209.0, rewards: 268 },
      { date: new Date("2013-01-04T23:00:00.000Z"), value: 152.01, rewards: 154 },
      { date: new Date("2013-01-01T23:00:00.000Z"), value: 52.02, rewards: 2 },
    ],
    rewardsSum: 424,
  },
  December: {
    transactions: [
      { date: new Date("2012-12-29T23:00:00.000Z"), value: 23.12, rewards: 0 },
      { date: new Date("2012-12-20T23:00:00.000Z"), value: 90.12, rewards: 40 },
      { date: new Date("2012-12-19T23:00:00.000Z"), value: 21.52, rewards: 0 },
      { date: new Date("2012-12-11T23:00:00.000Z"), value: 223.12, rewards: 296 },
    ],
    rewardsSum: 336,
  },
  November: {
    transactions: [
      { date: new Date("2012-11-28T23:00:00.000Z"), value: 150.12, rewards: 150 },
      { date: new Date("2012-11-19T23:00:00.000Z"), value: 100.02, rewards: 50 },
    ],
    rewardsSum: 200,
  },
}

const mockUseRewardsPage = vi.fn()
vi.mock("./useRewardsPage", () => ({
  useRewardsPage: () => mockUseRewardsPage(),
}))

describe("RewardsPage", () => {
  beforeEach(() => {
    vi.clearAllMocks()
    mockUseRewardsPage.mockReturnValue({
      user: { name: "John" },
      loading: false,
      rewards: mockData,
      total: 960,
    })
  })
  it("renders loader when loading is true", () => {
    mockUseRewardsPage.mockReturnValue({
      user: { name: "John" },
      loading: true,
      rewards: null,
      total: 0,
    })

    const {unmount, container} = render(<RewardsPage />)

    expect(container).toMatchSnapshot()
    unmount()
  })
  it("renders and match snapshot", () => {
    const {unmount, container} = render(<RewardsPage />)

    expect(container).toMatchSnapshot()
    unmount()
  })
  it("renders user name when user is present", () => {
    const {unmount} = render(<RewardsPage />)

    expect(screen.getByText("Welcome John!")).toBeTruthy()
    unmount()
  })
  it("renders total points", () => {
    const {unmount} = render(<RewardsPage />)

    expect(screen.getByText("Total points: 960")).toBeTruthy()
    unmount()
  })
})
