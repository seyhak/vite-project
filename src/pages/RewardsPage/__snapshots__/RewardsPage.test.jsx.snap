// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RewardsPage > renders and match snapshot 1`] = `
<div>
  <div
    class="RewardsPage"
  >
    <h3>
      Welcome John!
    </h3>
    <div
      class="rewards"
    >
      <div
        class="card rewards-group"
      >
        <h4
          class="month"
        >
          January
        </h4>
        <div
          class="month-rewards-wrapper"
        >
          <div
            class="Reward"
          >
            <div>
              <h6
                class="date"
              >
                Sun Jan 06 2013
              </h6>
              <h5
                class="value"
              >
                209
              </h5>
            </div>
            <div
              class="reward"
            >
              <h5>
                268
              </h5>
            </div>
          </div>
          <div
            class="Reward"
          >
            <div>
              <h6
                class="date"
              >
                Sat Jan 05 2013
              </h6>
              <h5
                class="value"
              >
                152.01
              </h5>
            </div>
            <div
              class="reward"
            >
              <h5>
                154
              </h5>
            </div>
          </div>
          <div
            class="Reward"
          >
            <div>
              <h6
                class="date"
              >
                Wed Jan 02 2013
              </h6>
              <h5
                class="value"
              >
                52.02
              </h5>
            </div>
            <div
              class="reward"
            >
              <h5>
                2
              </h5>
            </div>
          </div>
        </div>
        <h4
          class="sum"
        >
          Your reward points for the month: 424
        </h4>
      </div>
      <div
        class="card rewards-group"
      >
        <h4
          class="month"
        >
          December
        </h4>
        <div
          class="month-rewards-wrapper"
        >
          <div
            class="Reward"
          >
            <div>
              <h6
                class="date"
              >
                Sun Dec 30 2012
              </h6>
              <h5
                class="value"
              >
                23.12
              </h5>
            </div>
            <div
              class="reward"
            >
              <h5>
                0
              </h5>
            </div>
          </div>
          <div
            class="Reward"
          >
            <div>
              <h6
                class="date"
              >
                Fri Dec 21 2012
              </h6>
              <h5
                class="value"
              >
                90.12
              </h5>
            </div>
            <div
              class="reward"
            >
              <h5>
                40
              </h5>
            </div>
          </div>
          <div
            class="Reward"
          >
            <div>
              <h6
                class="date"
              >
                Thu Dec 20 2012
              </h6>
              <h5
                class="value"
              >
                21.52
              </h5>
            </div>
            <div
              class="reward"
            >
              <h5>
                0
              </h5>
            </div>
          </div>
          <div
            class="Reward"
          >
            <div>
              <h6
                class="date"
              >
                Wed Dec 12 2012
              </h6>
              <h5
                class="value"
              >
                223.12
              </h5>
            </div>
            <div
              class="reward"
            >
              <h5>
                296
              </h5>
            </div>
          </div>
        </div>
        <h4
          class="sum"
        >
          Your reward points for the month: 336
        </h4>
      </div>
      <div
        class="card rewards-group"
      >
        <h4
          class="month"
        >
          November
        </h4>
        <div
          class="month-rewards-wrapper"
        >
          <div
            class="Reward"
          >
            <div>
              <h6
                class="date"
              >
                Thu Nov 29 2012
              </h6>
              <h5
                class="value"
              >
                150.12
              </h5>
            </div>
            <div
              class="reward"
            >
              <h5>
                150
              </h5>
            </div>
          </div>
          <div
            class="Reward"
          >
            <div>
              <h6
                class="date"
              >
                Tue Nov 20 2012
              </h6>
              <h5
                class="value"
              >
                100.02
              </h5>
            </div>
            <div
              class="reward"
            >
              <h5>
                50
              </h5>
            </div>
          </div>
        </div>
        <h4
          class="sum"
        >
          Your reward points for the month: 200
        </h4>
      </div>
    </div>
    <div
      class="total card"
    >
      <h4
        class="sum"
      >
        Total points: 960
      </h4>
    </div>
  </div>
</div>
`;

exports[`RewardsPage > renders loader when loading is true 1`] = `
<div>
  <div
    class="RewardsPage"
  >
    <div
      class="Loader"
    />
  </div>
</div>
`;
